url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/LevelScene.java

```diff
@@ -0,0 +1,1076 @@
+    final private List<Sprite> spritesToAdd = new ArrayList<Sprite>();
+    final private List<Sprite> spritesToRemove = new ArrayList<Sprite>();
+
+    public Level level;
+    public Mario mario;
+    public float xCam, yCam, xCamO, yCamO;
+
+    public int tick;
+
+    public boolean paused = false;
+    public int startTime = 0;
+    public int timeLeft;
+    private int width;
+    private int height;
+
+    public boolean visualization = false;
+
+    final private int rows = Environment.HalfObsHeight*2+1;
+    final private int cols = Environment.HalfObsWidth*2+1;
+    final private int[] serializedLevelScene = new int[rows * cols];
+    final private int[] serializedEnemies = new int[rows * cols];
+    final private int[] serializedMergedObservation = new int[rows * cols];
+
+    final private byte[][] levelSceneZ = new byte[rows][cols];
+    final private byte[][] enemiesZ = new byte[rows][cols];
+    final private byte[][] mergedZ = new byte[rows][cols];
+
+    final private List<Float> enemiesFloatsList = new ArrayList<Float>();
+    final private float[] marioFloatPos = new float[2];
+    final private int[] marioState = new int[12];
+    private int numberOfHiddenCoinsGained;
+
+//    public int getTimeLimit() {  return timeLimit; }
+    public void setTimeLimit(int timeLimit) {  this.timeLimit = timeLimit; }
+
+    private int timeLimit = 200;
+
+    //    private Recorder recorder = new Recorder();
+    //    private Replayer replayer = null;
+
+    private long levelSeed;
+    private int levelType;
+    private int levelDifficulty;
+    private int levelLength;
+    private int levelHeight;
+    public static int killedCreaturesTotal;
+    public static int killedCreaturesByFireBall;
+    public static int killedCreaturesByStomp;
+    public static int killedCreaturesByShell;
+
+    private int[] args; //passed to reset method. ATTENTION: not cloned.
+
+    public LevelScene(long seed, int levelDifficulty, int type, int levelLength, int levelHeight, int timeLimit, int visualization)
+    {
+        this.levelSeed = seed;
+        this.levelDifficulty = levelDifficulty;
+        this.levelType = type;
+        this.levelLength = levelLength;
+        this.levelHeight = levelHeight;
+        this.setTimeLimit(timeLimit);
+        this.visualization = visualization == 1;
+        killedCreaturesTotal = 0;
+        killedCreaturesByFireBall = 0;
+        killedCreaturesByStomp = 0;
+        killedCreaturesByShell = 0;
+
+        // moved from init;
+        try
+        {
+//            System.out.println("Java::LevelScene: loading tiles.dat...");
+//            System.out.println("LS: System.getProperty(\"user.dir()\") = " + System.getProperty("user.dir"));
+
+            //Level.loadBehaviors(new DataInputStream(LevelScene.class.getResourceAsStream("C:\\tiles.dat")));
+            Level.loadBehaviors(new DataInputStream(LevelScene.class.getResourceAsStream("resources/tiles.dat")));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *DataInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when Level.loadBehaviors throws an exception. Ensure the object returned by *DataInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).